hbase-client/src/main/java/org/apache/hadoop/hbase/client/AsyncTableRegionLocatorImpl.java

```diff
@@ -65,13 +65,17 @@ public CompletableFuture<List<HRegionLocation>> getAllRegionLocations() {
       CompletableFuture<List<HRegionLocation>> future = new CompletableFuture<>();
       addListener(ClientMetaTableAccessor.getTableHRegionLocations(conn
           .getTable(TableName.META_TABLE_NAME), tableName), (locs, error) -> {
-        if (error != null) {
-          future.completeExceptionally(error);
-          return;
-        }
-        locs.forEach(loc -> conn.getLocator().getNonMetaRegionLocator().addLocationToCache(loc));
-        future.complete(locs);
-      });
+          if (error != null) {
+            future.completeExceptionally(error);
+            return;
+          }
+          try {
+            locs.forEach(loc -> conn.getLocator()
+              .getNonMetaRegionLocator().addLocationToCache(loc));
```

Comment:
Will addLocationToCache throw exception? I'm not sure.

Reply:
Thanks for reviewing Duo.

Not likely. But if it does, the future will hang up forever. So maybe we should catch the potential exception here.

Thanks.

Reply:
There are no explicitly thrown exceptions but we've certainly leaked accidental bugs in the past some places, like NPEs, etc. If we are going to have an extra future in the callback, we need to make sure it's always completed even in unforseeen circumstances. Alternative we could remove the extra future.

Reply:
At least I do not think we should fail the original future when we fail to add them to cache?

Reply:
Ok, I was thinking we should try to have a similar behavior to the blocking client, which would throw an exception if the cache is unable to be loaded. I think many people will use this method as a way to fill the cache on process startup, so it's not great for them if it silently fails to load the cache. But if you don't like it maybe we can put a separate jira in to add a `loadRegionCache()` method if necessary.

Reply:
We also use it in other place, in our own code, for example, the getStartEndKeys is implemented based on getAllRegionLocations. I do not think it is a good idea to fail this operation when we fail to add the result to cache? The caller will be confused...

I think adding a special method to fill the cache is a possible solution. You can start a discussion thread on the dev and user list to discuss it.

Thanks.

Reply:
Ok sounds good. In that case @frostruan I would probably just remove the extra futures altogether:

```java
CompletableFuture<List<HRegionLocation>> future = ClientMetaTableAccessor.getTableHRegionLocations(conn
          .getTable(TableName.META_TABLE_NAME), tableName);
addListener(future, (locs, error) -> {
 // cache
});
return future;
```

No need to handle errors since they're already handled in addListener and whenComplete will not block the client waiting on the original future.

Sorry for the back and forth

Reply:
+1

Reply:
> We also use it in other place, in our own code, for example, the getStartEndKeys is implemented based on getAllRegionLocations. I do not think it is a good idea to fail this operation when we fail to add the result to cache? The caller will be confused...
> 
> I think adding a special method to fill the cache is a possible solution. You can start a discussion thread on the dev and user list to discuss it.
> 
> Thanks.

a little awkward, for some weird unknown reasons, I can't send emails to the dev list ...

But it's fine. I think the solution above is great. I'll address it. Thanks all for the reviewing and discussion. @bbeaudreault @Apache9